src/main/scala/org/constellation/consensus/ResolutionService.scala

```diff
@@ -0,0 +1,109 @@
+import akka.util.Timeout
+import org.constellation.Data
+import org.constellation.primitives.{APIBroadcast, Schema}
+import org.constellation.primitives.Schema.{CheckpointBlock, CheckpointCacheData, SignedObservationEdgeCache}
+
+import scala.annotation.tailrec
+import scala.concurrent.{ExecutionContext, Future}
+
+object ResolutionService {
+  implicit val timeout: Timeout = Timeout(5, TimeUnit.SECONDS)
+
+  case class ResolutionStatus(
+                               cb: CheckpointCacheData,
+                               resolvedParents: Seq[(String, Option[SignedObservationEdgeCache])],
+                               unresolvedParents: Seq[(String, Option[SignedObservationEdgeCache])]
+                             )
+
+  /**
+    * TODO: Need to include signatories ABOVE this checkpoint block later in the case of signature decay. Add to SignedObservationEdgeCache
+    *
+    * @param dao
+    * @param cb
+    * @return
+    */
+  def downloadAncestry(dao: Data, cb: CheckpointBlock) = {
+    val observers: Set[Schema.Id] = cb.signatures.map {
+      _.toId
+    }.toSet
+    dao.queryMissingResolutionData(cb.baseHash, observers)
+  }
+
+  /**
+    *
+    * @param dao
+    * @param unresolvedChildren
+    */
+  def reprocessUnresolvedChildren(dao: Data, unresolvedChildren: Seq[CheckpointBlock])(implicit executionContext: ExecutionContext) = unresolvedChildren.foreach { unresolvedChild =>
+    EdgeProcessor.handleCheckpoint(unresolvedChild, dao, true) //Todo log exception add retry
+  }
+
+  /**
+    *
+    * @param dao
+    * @param cb
+    */
+  def reprocessUnresolvedParents(dao: Data, cb: CheckpointBlock): (String, Option[SignedObservationEdgeCache]) => Unit = {//Todo test me
+    case (missingParentHash: String, queryResult: Option[SignedObservationEdgeCache]) =>
+      if (queryResult.isEmpty) downloadAncestry(dao, cb)
+      dao.markParentsUnresolved(missingParentHash, cb)
+  }
+
+  /**
+    * If this newly resolved checkpoint block was a previously unresolved parent, reprocess the child
+    *
+    * @param dao
+    * @param children
+    * @param executionContext
+    * @return
+    */
+  def isUnresolved(dao: Data, children: Seq[CheckpointBlock])(implicit executionContext: ExecutionContext): Seq[CheckpointBlock] = {
+    children.filter { child =>
+      dao.dbActor.getSignedObservationEdgeCache(child.baseHash).exists(!_.resolved)
+    }
+  }
+
+  /**
+    *
+    * @param dao
+    * @param cb
+    * @return
+    */
+  def partitionByParentsResolved(dao: Data, cb: CheckpointBlock)(implicit executionContext: ExecutionContext):
+  (Seq[(String, Option[SignedObservationEdgeCache])], Seq[(String, Option[SignedObservationEdgeCache])]) = {
+    val parentCache = cb.checkpoint.edge.parentHashes.map { hash => (hash, dao.dbActor.getSignedObservationEdgeCache(hash)) }
+    val (unresolvedParents, resolvedParents) = parentCache.partition { case (hash, queryResult) => queryResult.exists(_.resolved) }
+    (unresolvedParents, resolvedParents)
```

Comment:
also, if you're just going to return the tuple, there's no need to decompose and recompose it. I'd just do ```val partitionParents = ...```
and return that, or just return the expression.

Reply:
good point